File: apps/tv/modules/valence-glass/ios/ValenceGlassView.swift
import ExpoModulesCore
import UIKit

/// Draws the system's Liquid Glass behind whatever it holds — refracting and tinting what lies behind
/// it and catching the light along its edge — where the television has it, from tvOS 26, and Apple's
/// dark frosted blur where it does not. The glass is kept beneath the views React Native puts in, and
/// rounded to the given radius, or into a capsule where the radius is more than half its height.
final class ValenceGlassView: ExpoView {
  var cornerRadius: CGFloat = 0 {
    didSet { setNeedsLayout() }
  }

  private let glass: UIVisualEffectView = {
    if #available(tvOS 26.0, iOS 26.0, *) {
      return UIVisualEffectView(effect: UIGlassEffect())
    }

    return UIVisualEffectView(effect: UIBlurEffect(style: .dark))
  }()

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    glass.isUserInteractionEnabled = false
    glass.clipsToBounds = true
    glass.layer.zPosition = -1
    addSubview(glass)
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    glass.frame = bounds
    glass.layer.cornerRadius = min(cornerRadius, bounds.height / 2)
    glass.layer.cornerCurve = .continuous
  }
}

File: apps/tv/modules/valence-glass/android/src/main/java/app/valence/tv/glass/ValenceGlassView.kt
package app.valence.tv.glass

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView

private val PANE = Color.argb(204, 28, 28, 30)

private val EDGE = Color.argb(46, 255, 255, 255)

/**
 * A dark pane behind whatever it holds, standing in for Apple's Liquid Glass: Android cannot blur
 * what lies behind a view, so this is the frosted look's colour and its thin catch of light along
 * the edge, rounded to the given radius, or into a capsule where that is more than half its height.
 */
class ValenceGlassView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  var cornerRadius = 0f
    set(value) {
      field = value
      shape()
    }

  private val density = resources.displayMetrics.density

  private val pane = GradientDrawable().apply {
    setColor(PANE)
    setStroke(density.toInt().coerceAtLeast(1), EDGE)
  }

  init {
    background = pane
  }

  override fun onSizeChanged(width: Int, height: Int, oldWidth: Int, oldHeight: Int) {
    super.onSizeChanged(width, height, oldWidth, oldHeight)
    shape()
  }

  /** Rounds the pane to its radius, or into a capsule where that is more than half its height. */
  private fun shape() {
    pane.cornerRadius = (cornerRadius * density).coerceAtMost(height / 2f)
  }
}
